Use Actions for CI and publishing

bluekeyes opened the pull request.

bluekeyes committed:
Use Actions for CI and publishing

bluekeyes committed:
Fix yarn commands, use Go 1.16

bluekeyes committed:
Fix GOPATH, add caching

bluekeyes committed:
Fix typo in cache keys

bluekeyes committed:
Fix cache keys, update go.sum for Go 1.16

bluekeyes committed:
Checkout full history so version computation works

bluekeyes committed:
Add publishing commands

bluekeyes committed:
Fix artifacts path

bluekeyes committed:
Remove CircleCI configuration

bluekeyes commented:
No idea if the release and publishing stuff will work, as it's kind of hard to test on a branch without making an actual release. Any ideas for testing this beyond careful review and then making point-releases-of-shame if it doesn't work?

bluekeyes commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,141 @@
+        UPLOAD_URL: ${{ steps.find-release.outputs.upload_url }}
+      run: |
+        dist_file=$(./godelw artifacts dist)
+        curl -sS --fail -X POST \
+          -H "Authorization: Bearer ${GITHUB_TOKEN}" \
+          -H "Content-Type: application/gzip" \
+          "${UPLOAD_URL}?name=$(basename ${dist_file})" \
+          --data-binary @${dist_file}
```
I feel kind of bad just doing this with `curl` but the semi-official action to upload release assets is archived/deprecated and most community ones I found seem to also want to create the release. I guess a lot of people want to push a tag and have that automate the release creation instead of creating the release as a way to create the tag.

Another option is to write a new publisher plugin for Godel that can do this and the previous step. That's probably the best long-term option, but I wasn't ready to write that.

I know @jamestoyer likes `goreleaser`, but it looked like it also handles a bunch of the build process that we currently use Godel for and also works on the "push a tag to create a release" model.

derekjobst replied:
`goreleaser` looks nice. Autorelease Bot also follows this workflow of Create Release/Tag → Run CI to add artifacts/dists to the release, right?

bluekeyes replied:
Yeah, Autorelease does create a tag before the release and then the tag creation triggers CI, but most (all?) of these projects don't actually attach artifacts to the release, but publish to other systems.

I was mainly flagging the workflow differences for people creating releases. Our current process is:

1. Publish a GitHub Release with release notes and a tag that does not exist yet
2. GitHub creates the tag
3. The tag triggers CI
4. CI publishes/upload artifacts

It seems like many of the open-source tools I found assume this process:

1. Locally create a tag and push it to GitHub
2. The tag triggers CI/automation
3. The automation creates a GitHub Release for the the existing tag
4. CI publishes/uploads artifacts

jamestoyer replied:
A lot of OSS tooling will also be fine with an existing release. They will just update that. Generally, I much prefer using the `release` event to trigger the actual building of a release artifact

jmcampanini commented:
this looks pretty dang cool! i defer to @jamestoyer on what to use. i can see a godel plugin being useful here.

bluekeyes commented:
I expect the Godel plugin will exist soon as part of moving the Godel ecosystem itself to publish to GitHub Releases, so we might just be ahead of the curve there.

derekjobst commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,141 @@
+        curl -sS --fail -H "Authorization: Bearer ${GITHUB_TOKEN}" "${GITHUB_API_URL}/repos/${GITHUB_REPOSITORY}/releases/tags/${GITHUB_REF#refs/tags/}" \
+          | jq .upload_url \
+          | sed 's/{.*}$//' \
+          | xargs -0 printf "::set-output name=upload_url::%s\n"
```
This feature of github actions is pretty cool! [Docs](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idoutputs) if anyone else wants to read more.

derekjobst reviewed (approved):
Excited for this!

nmiyake commented:
Will likely discuss in our next sync, but we've actually already rolled out OSS publishing for the godel ecosystem using GitHub actions and have excavated it out across OSS. Here's an example: https://github.com/palantir/godel-generate-plugin/blob/master/.github/workflows/publish-godel-artifacts.yml

asvoboda commented:
> Will likely discuss in our next sync, but we've actually already rolled out OSS publishing for the godel ecosystem using GitHub actions and have excavated it out across OSS. Here's an example: https://github.com/palantir/godel-generate-plugin/blob/master/.github/workflows/publish-godel-artifacts.yml

In the case of policy-bot, we really only publish a container and don't rely on the resulting tarball at all in prod. However, we recognize this is useful to some of the folks externally who use policy bot. This PR is intended to replace all of our circleci usage with actions, not just the publishing aspect.

bluekeyes committed:
Use Godel to upload GH release assets

asvoboda reviewed (approved):
Lets try it out!

bluekeyes committed:
Use release event as publish trigger

bluekeyes commented:
Ok, I added the secrets by hand and updated the workflow to use `release` events, so I think I'm ready to flip the required status checks and try this out.

asvoboda commented:
🎉  👍
